cns/ipampool/v2/monitor.go

```diff
@@ -0,0 +1,167 @@
+package v2
+
+import (
+	"context"
+	"math"
+	"sync"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/crd/clustersubnetstate/api/v1alpha1"
+	"github.com/Azure/azure-container-networking/crd/nodenetworkconfig/api/v1alpha"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	// DefaultMaxIPs default maximum allocatable IPs on a k8s Node.
+	DefaultMaxIPs = 250
+	// fieldManager is the field manager used when patching the NodeNetworkConfig.
+	fieldManager = "azure-cns"
+)
+
+type nodeNetworkConfigSpecUpdater interface {
+	PatchSpec(context.Context, *v1alpha.NodeNetworkConfigSpec, string) (*v1alpha.NodeNetworkConfig, error)
+}
+
+type ipStateStore interface {
+	GetPendingReleaseIPConfigs() []cns.IPConfigurationStatus
+	MarkNIPsPendingRelease(n int) (map[string]cns.IPConfigurationStatus, error)
+}
+
+type scaler struct {
+	batch     int64
+	buffer    float64
+	exhausted bool
+	max       int64
+}
+
+type Monitor struct {
+	z            *zap.Logger
+	scaler       scaler
+	nnccli       nodeNetworkConfigSpecUpdater
+	store        ipStateStore
+	demand       int64
+	request      int64
+	demandSource <-chan int
+	cssSource    <-chan v1alpha1.ClusterSubnetState
+	nncSource    <-chan v1alpha.NodeNetworkConfig
+	started      chan interface{}
+	once         sync.Once
+}
+
+func NewMonitor(z *zap.Logger, store ipStateStore, nnccli nodeNetworkConfigSpecUpdater, demandSource <-chan int, nncSource <-chan v1alpha.NodeNetworkConfig, cssSource <-chan v1alpha1.ClusterSubnetState) *Monitor { //nolint:lll // it's fine
+	return &Monitor{
+		z:            z.With(zap.String("component", "ipam-pool-monitor")),
+		store:        store,
+		nnccli:       nnccli,
+		demandSource: demandSource,
+		cssSource:    cssSource,
+		nncSource:    nncSource,
+		started:      make(chan interface{}),
+	}
+}
+
+// Start begins the Monitor's pool reconcile loop.
+// On first run, it will block until a NodeNetworkConfig is received (through a call to Update()).
+// Subsequently, it will run run once per RefreshDelay and attempt to re-reconcile the pool.
+func (pm *Monitor) Start(ctx context.Context) error {
+	pm.z.Debug("starting")
+	for {
+		// proceed when things happen:
+		select {
+		case <-ctx.Done(): // calling context has closed, we'll exit.
+			return errors.Wrap(ctx.Err(), "pool monitor context closed")
+		case demand := <-pm.demandSource: // updated demand for IPs, recalculate request
+			pm.demand = int64(demand)
+			pm.z.Info("demand update", zap.Int64("demand", pm.demand))
+		case css := <-pm.cssSource: // received an updated ClusterSubnetState, recalculate request
+			pm.scaler.exhausted = css.Status.Exhausted
+			pm.z.Info("exhaustion update", zap.Bool("exhausted", pm.scaler.exhausted))
+		case nnc := <-pm.nncSource: // received a new NodeNetworkConfig, extract the data from it and recalculate request
```

Comment:
along similar lines as my comment above, I'd eventually like to get us to a point where we're not operating on NNC specs/statuses in this package but instead the nncSource contains some domain type `IPState` instead of a `v1alpha.NodeNetworkConfigStatus`

Reply:
will do once I can delete the v1 Monitor...until then it complicates the NNC reconciler too much